# <font color='#2B17B0'>CIENCIA DE DATOS APLICADA A FINANZAS</font>
Sesión 9, 2024-11-05<br>
Carlos Crespo Elizondo, MSF<br>
EGADE BUSINESS SCHOOL

## <font color='#2B17B0'>COMENTARIOS INICIALES</font>

### Contacto

email: carlos@oncedos.com

### Asesorías

Bajo previa cita con un día de anticipación:

### Github:  Clonar repositorio de la clase
URL del repositorio:<br>
https://github.com/carlosdeoncedos/CienciaDatosFinanzas2024_11<br>

Para verificar si GitHub está instaldo corre el siguiente comando desde la ventana de terminal:
`git --version`.  Si está instalado, arrojará la versión instalada.



#### Instalar GitHub
En caso de no tener Github instalado:

##### **MAC OS:**
1. Abre la Terminal en tu Mac.
1. Ejecuta el siguiente comando para instalar Homebrew:<br>
    `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
1. Ejecuta el siguiente comando para instalar Git usando Homebrew:
    `brew install git`

##### **WINDOWS**:
1. Ve al sitio web oficial de Git: https://git-scm.com/download/win
1. La descarga del instalador de Git para Windows debería comenzar automáticamente. Si no es así, haz clic en el enlace "click here to download manually" (haz clic aquí para descargar manualmente) en la página.
1. Ejecuta el archivo de instalación que descargaste en el Paso 1.
1. Selecciona las opciones de instalación que prefieras y sigue las instrucciones del instalador. Si no estás seguro de qué opciones elegir, las predeterminadas suelen ser adecuadas para la mayoría de los usuarios.
1. Haz clic en "Install" (Instalar) para comenzar la instalación.
1. Una vez que se complete la instalación, Git estará instalado en tu computadora con Windows.

#### **CREAR USUARIO**
1. Entra a la página https://github.com
1. En la página principal de GitHub, localiza el formulario de registro.
1. Ingresa un nombre de usuario único. GitHub verificará si el nombre de usuario está disponible.
1. Ingresa una dirección de correo electrónico válida. Asegúrate de que sea una dirección de correo electrónico a la que tengas acceso, ya que GitHub enviará un correo electrónico de verificación.
1. Crea una contraseña segura siguiendo las pautas de GitHub. La contraseña debe tener al menos 8 caracteres y debe incluir letras mayúsculas y minúsculas, números y símbolos.
1. Haz clic en el botón "Sign up for GitHub" (Regístrate en GitHub) para continuar.
1. GitHub te enviará un correo electrónico de verificación a la dirección de correo electrónico que proporcionaste.
1. Abre el correo electrónico y haz clic en el enlace de verificación para confirmar tu dirección de correo electrónico y continuar con el proceso de registro.
1. GitHub te mostrará diferentes planes disponibles. Elige el plan gratuito y haz clic en "Continue". El plan gratuito te permitirá crear repositorios públicos ilimitados y colaborar en repositorios privados.
1. GitHub te ofrecerá la opción de personalizar tu experiencia proporcionando información adicional sobre tus intereses y preferencias. Puedes completar esta información si lo deseas, pero no es obligatorio.
1. Una vez que hayas proporcionado la información que desees, haz clic en "Submit" (Enviar) o "Skip this step" (Omitir este paso) si prefieres no proporcionar esta información.

#### Clonar repositorio

1. Navega a la página principal del repositorio de la clase (Ver URL).
1. Haz clic en el botón verde "Code" cerca de la esquina superior derecha de la página.
1. En el menú desplegable, elige entre HTTPS o SSH en "Clone". Si no estás seguro de cuál usar, se recomienda HTTPS para la mayoría de los usuarios debido a su simplicidad y compatibilidad. Copia la URL proporcionada.
1. Abre tu terminal (en macOS o Linux) o Command Prompt (en Windows) en la computadora y ve al directorio donde deseas clonar el repositorio
1. Ejecutar el siguiente comando en tu terminal/Command Prompt (en Windows) para clonar el repositorio:<br>
    `git clone https://github.com/carlosdeoncedos/CienciaDatosFinanzas2024_11`

## <font color='#2B17B0'>LIBRERIAS</font>

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

sns.set_style('darkgrid')

## <font color='#2B17B0'>RETORNOS</font>

### Definición:

La mayoría de los estudios financieros se concentran en los retornos en lugar de los precios de los activos. Hay varias razones algunas de ellas son:
1. Para los inversionistas promedio, el rendimiento de un activo proporciona un resumen completo y sin escala de la oportunidad de inversión.
2. Las series de retornos son más fáciles de manejar que las series de precios debido a sus propiedades estadísticas más favorables.

> _El **retorno** es un indicador esencial que refleja la ganancia o pérdida en una inversión en comparación con la cantidad invertida inicialmente, y se expresa en forma de porcentaje o decimal. Este indicador nos permite evaluar el desempeño de las inversiones, comparar diversas opciones y tomar decisiones fundamentadas en la asignación de recursos. Alcanzar el equilibrio adecuado entre riesgo y retorno es crucial para desarrollar una estrategia de inversión exitosa._<br><br>
> _**IMPORTANTE**: Por lo general los rendimientos **más altos**  conllevan **mayores riesgos**._<br>

### Cálculo del retorno simple:

<br><br>
$$\Large R=\frac{P_1-P_0}{P_0}\rightarrow \frac{P_1}{P_0}-1$$

Ejemplo Precio de la acción ACME <br>
$P_{t_0}$: 50.00 <br>
$P_{t_1}$: 55.00

In [2]:
p0 = 50
p1 = 55
r = (p1 - p0) / p0
print(r)

0.1


In [3]:
r = (p1 / p0) - 1
print(r)


0.10000000000000009


## <font color='#2B17B0'>FUNCIONES, LIBRERIAS Y COMANDOS MAGICOS</font>

### CREAR FUNCIONES

Una función es un block de código que está diseñado a hacer una función en especial. Cuando queremos utilizar la función, la llamamos utilizando su nombre. <br><br>
En python las funciones se definenen utilizando `def` y siempre va seguida del nombre de la función seguido de un par de paréntesis.

#### CREAR FUNCION `retorno_simple`

La función debe:
* Tener dos parámetros `precio_inicial` y `precio_final`.
* Calcular el retorno simple dados los dos parámetros.
* Asegurarse que los argumentos (los valores asignados a los parámetros), sean del tipo numérico.
* En caso de no ser así, generar un error que diga `"Los parámetros, 'x' y 'y', deben ser valores numéricos (int o float)."`.

In [4]:
def retorno_simple(precio_inicial, precio_final):
    if not isinstance(precio_inicial, (int, float)) or not isinstance(precio_final, (int, float)):
        raise TypeError("Los parámetros 'precio_inicial' y/o 'precio_final' deben de ser valores numéricos")
    retorno = (precio_final / precio_inicial) - 1

    return retorno

In [5]:
retorno_simple(1.8, 2)

0.11111111111111116

In [6]:
isinstance(.1, (int, float))

True

### LIBRERIAS

La función `retorno_simple`, podemos utilizarla en otras libretas de Jupyter o archivos de python si la grabamos en un archivo de Python y la importamos como si fuera una librería como Pandas, Numpy, etc.

In [7]:
import herramientas_financieras as hf

In [11]:
hf.retorno_simple(1.8, 2)

0.11111111111111116

### COMANDOS MAGICOS

En Jupyter, existen instrucciones especiales independientes del código de Python que se utlizan para diferentes tareas. 

Algunos comandos mágicos:

<table>
    <tr>
        <th>Comando Mágico</th>
        <th>Descripción</th>
        <th>Código de Ejemplo</th>
    </tr>
    <tr>
        <td><code>%time</code></td>
        <td>Mide el tiempo de ejecución de una sola línea de código.</td>
        <td><code>%time sum(range(1000))</code></td>
    </tr>
    <tr>
        <td><code>%timeit</code></td>
        <td>Ejecuta la línea de código varias veces y calcula el tiempo promedio.</td>
        <td><code>%timeit sum(range(1000))</code></td>
    </tr>
    <tr>
        <td><code>%who</code></td>
        <td>Muestra todas las variables definidas en el espacio de trabajo actual.</td>
        <td><code>%who</code></td>
    </tr>
    <tr>
        <td><code>%matplotlib inline</code></td>
        <td>Hace que los gráficos de Matplotlib se muestren dentro del cuaderno.</td>
        <td><code>%matplotlib inline</code></td>
    </tr>
    <tr>
        <td><code>%ls</code></td>
        <td>Lista los archivos en el directorio actual, similar a <code>ls</code> en la terminal.</td>
        <td><code>%ls</code></td>
    </tr>
    <tr>
        <td><code>%run &lt;filename&gt;</code></td>
        <td>Ejecuta un script de Python desde un archivo.</td>
        <td><code>%run script.py</code></td>
    </tr>
    <tr>
        <td><code>%debug</code></td>
        <td>Entra en el modo de depuración interactivo en caso de que ocurra un error.</td>
        <td><code>%debug</code></td>
    </tr>
    <tr>
        <td><code>%load &lt;filename&gt;</code></td>
        <td>Carga el contenido de un archivo en la celda actual.</td>
        <td><code>%load script.py</code></td>
    </tr>
</table>


In [20]:
%ls

README.md                    __pycache__/
Sesion9.ipynb                herramientas_financieras.py


#### Comandos mágicos para cargar automáticamente cambios en un archivo externo

In [9]:
%load_ext autoreload
%autoreload 2

`%load_ext autoreload`: Este comando carga la extensión autoreload, que permite la recarga automática de módulos.

`%autoreload 2`: Activa la recarga automática de todos los módulos importados cada vez que se ejecuta una celda. Esto significa que cualquier cambio que hagas en un archivo externo de Python se reflejará inmediatamente en el cuaderno sin necesidad de reiniciar el kernel o volver a importar manualmente.

Desarrollar su docstring (utilizar ChatGPT) para la función `retorno_simple`

In [ ]:
pd.DataFrame(

## <font color='#2B17B0'>FLUJOS DE EFECTIVOS Y RETORNOS ENTRE MULTIPLES PERIODOS</font>

### Flujos de efectivo

> _Los flujos de efectivo representan el movimiento de dinero que se recibe o se gasta a lo largo de un período determinado. Estos flujos pueden ser entrantes (dinero recibido) o salientes (dinero gastado), y su ocurrencia puede ser regular o esporádica._
<br><br>
> _Visualizar el flujo de efectivo en un gráfico facilita la comprensión de los datos financieros complejos, mejorando así la toma de decisiones. Los gráficos muestran la dirección de los flujos (entrantes o salientes), las cantidades de dinero, el número de períodos y la frecuencia con la que ocurren._

![Image](https://dl.dropboxusercontent.com/scl/fi/k4vrvw71s9kw5j16hzq17/flujos_periodo.png?rlkey=yu3c96xo5x3izv0sno4wi4h7n&dl=1)

### Calcular el retorno de los flujos de $t_0$ a $t_1$ y de $t_1$ a $t_2$

![Image](https://dl.dropboxusercontent.com/scl/fi/7ccrpxhi9vqt1mfcpfbuu/retorno_periodo.png?rlkey=r5c8wno09mxna74q918m46mwl&dl=1)

In [12]:
r01 = hf.retorno_simple(140, 189)
print(r01)

0.3500000000000001


In [13]:
r12 = hf.retorno_simple(189, 195)
print(r12)

0.031746031746031855


### Calcular el retorno múltiple o compuesto de $t_0$ a $t_2$

> *El retorno de múltiples periodos es la medida acumulativa de la ganancia o pérdida de una inversión a lo largo de varios intervalos de tiempo consecutivos, reflejando el efecto compuesto de los retornos en cada periodo.*

![Image](https://dl.dropboxusercontent.com/scl/fi/89iafhshon8fng5415kci/retorno_multiples_periodos.png?rlkey=d6wluqutujqsl1wj24gjdkg3f&dl=1)

In [14]:
r02 = hf.retorno_simple(140, 195)
print(r02)

0.3928571428571428


Si calculamos el retorno utilizando los valores en $t_0$ y $t_2$, estamos calculando un retorno total, PERO:
* Cuando calculamos el retorno de múltiples períodos de esta manera, estamos capturando solo el cambio entre el inicio y el final, pero **no estamos considerando cómo los rendimientos intermedios afectan el valor acumulado**.
* El efecto del interés compuesto es crucial porque **cada período de rendimiento afecta el valor total de la inversión**, lo que significa que las ganancias o pérdidas en cada paso se acumulan. Ignorar esto da una visión incompleta de la evolución real de la inversión.
* Los retornos simples de múltiples períodos **no capturan correctamente la volatilidad**. En períodos con grandes fluctuaciones (por ejemplo, un aumento grande seguido de una caída), el retorno simple puede dar una impresión engañosa.
* Al usar retornos simples, puede ser difícil comparar entre distintos horizontes de tiempo (por ejemplo, entre retornos anuales y mensuales) ya que el retorno simple de múltiples períodos no permite una conversión directa a una tasa anual o mensual.
* En términos prácticos, los inversionistas suelen estar interesados en el valor futuro de su inversión, que depende del crecimiento compuesto. Al usar solo el retorno simple, **se pierde la capacidad de medir cómo crece realmente la riqueza en el tiempo**.
* Esto es importante porque **el rendimiento compuesto es una medida más realista del potencial de crecimiento**, mientras que el retorno simple puede presentar una visión simplificada que no se alinea con el comportamiento real de la inversión.

$$
R_{\text{total}} = \left[ \prod_{t=0}^{n} (1 + R_t) \right] - 1
$$

$\large R_{t_0,t_2}=(1+R_{t_0,t_1})(1+R_{t_1,t_2})-1$
<br><br>
$\large R_{t_0,t_2}=(1+0.35)(1+0.032)-1$ <br><br>

In [15]:
print(r01, r12)

0.3500000000000001 0.031746031746031855


In [16]:
(1 + r01) * (1 + r12) - 1

0.392857142857143

#### CREAR FUNCION `retorno_compuesto`

La función debe:
* Mismas características que la función `retorno_simple`.
* Los datos deben de ser ingresados en una lista.

In [41]:
def retorno_compuesto(valores):
    if len(valores) < 2:
        raise ValueError('A la lista le faltan más valores')

    # Calcular el retorno de cada periodo y acumular el retorno compuesto
    retorno_compuesto = 1

    for i in range(1, len(valores)):
        precio_inicial = valores[i - 1]
        precio_final = valores[i]

        if not isinstance(precio_inicial, (int, float)) or not isinstance(precio_final, (int, float)):
            raise TypeError("Los parámetros 'precio_inicial' y/o 'precio_final' deben de ser valores numéricos")
    
        retorno_periodo = (precio_final / precio_inicial) - 1
        retorno_compuesto *= (1 + retorno_periodo)

    retorno_compuesto = retorno_compuesto - 1

    return retorno_compuesto

In [46]:
retorno_compuesto(['x', 189, 195])

TypeError: Los parámetros 'precio_inicial' y/o 'precio_final' deben de ser valores numéricos

In [45]:
2 * 4
3 * 4
5 * 4

20

In [20]:
lista_for_loop = [2, 3, 4, 5, 6,7 ,8, 9]

for elemento in lista_for_loop:
    multiplicacion = elemento * 2
    print(elemento, multiplicacion)

2 4
3 6
4 8
5 10
6 12
7 14
8 16
9 18


In [21]:
# lista_for_loop = [2, 3, 4, 5, 6,7 ,8, 9]

for elemento in range(2, 10):
    multiplicacion = elemento * 2
    print(elemento, multiplicacion)

2 4
3 6
4 8
5 10
6 12
7 14
8 16
9 18


In [22]:
len(lista_for_loop)

8

In [26]:
lista_for_loop[7]

9

In [27]:
lista_for_loop[-1]

9

In [28]:
lista_for_loop[-2]

8

In [29]:
lista_for_loop[-8]

2

In [19]:
retorno_compuesto([1, 2])

## <font color='#2B17B0'>RETORNOS ANUALIZADOS</font>

Anualizar retornos es fundamental para realizar comparaciones exactas y equitativas de la rentabilidad de distintas inversiones, especialmente cuando los periodos de inversión difieren.<br><br>
Anualizar permite responder preguntas como: "¿Cuál sería el rendimiento anual si este rendimiento mensual se mantuviera constante durante todo el año?" Al expresar todos los retornos en una misma base temporal (anual), podemos comparar inversiones o analizar el crecimiento en términos consistentes y uniformes.<br><br>

### De retorno mensual a retorno anual

Tenemos una inversión que nos da un retorno mensual $R_m$. ¿Cual será el _Valor final_ de la inversión dentro de un año?

$$ ValorFinal = ValorInicial(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)(1+R_m)$$
$$\Large ValorFinal = VarlosInicial(1+R_m)^{12}$$
Ahora queremos encontrar un Retorno anual que sea equivalente al retorno compuesto menusal
$$\Large ValorInicial(1+R_a)=ValorInicial(1+R_m)^{12}$$
Al dividir la ecuación entre $ValorInicial$ y despejando para $R_a$ tenemos:
$$
\begin{align*}
\Large 1+R_a &= \Large(1+R_m)^{12}\\
\Large R_a &= \Large(1+R_m)^{12}-1\\
\end{align*}
$$

Tenemos una inversión que nos generó 3% en un mes, cual es su tasa anual?

$$
\begin{align*}
\Large R_{a} &= \Large(1 + R_{P})^{12} - 1 \\
\Large R_{a} &= \Large(1 + 0.03)^{12} - 1 \\
\Large R_{a} &= \Large1.4257 - 1 \\
\Large R_{a} &= \Large0.4257
\end{align*}
$$



Donde<br>
* $R_a$: Retorno anualizado<br>
* $R_P$: Retorno del periodos<br>

Siguiendo la misma lógica una inversión que nos generó 17% en un 6 meses, cual es su tasa anual?

$$
\begin{align*}
\Large (1 + R_{a}) &= \Large(1 + R_{s})^{?} \\
\Large R_{a} &= \Large(1 + R_{s})^{?} - 1 \\
\Large R_{a} &= \Large(1 + 0.17)^{?} - 1 \\
\Large R_{a} &= \Large? - 1 \\
\Large R_{a} &= \Large?
\end{align*}
$$

Siguiendo estos dos ejemplos, podemos generalizar y llegar a la siguiente fórmula:
$$\Large R_{a}=(1+R_{P})^{PeriodosPorAño}-1$$

Anualizar las siguientes dos tasas:
1. Tasa trimestral de 5%
2. Tasa diaria de 0.1%

$$\Large R_{a}=(1+R_{P})^{\frac{PeriodosPorAño}{NúmeroDePeriodos}}-1$$
Donde<br>
* $R_a$: Retorno anualizado<br>
* $R_P$: Retorno del periodos<br>
* $PeriodosPorAño$:  Números de periodos presentes en un año<br>
* $NúmeroDePeriodos$:  Números de periodos que representa el $R_p$

Tenemos una inversión que nos generó un retorno de 1% en 3 dias. ¿Cuál sería su tasa anual equivalente?

|Frecuencia|Fórmula|
|:-:|:-:|
|Base|$$(1+R_p)^{\frac{PeriodosPorAño}{NúmeroPeriodos}}$$
|Diaria*|$$(1+R_p)^{365}-1$$|
|Diaria*|$$(1+R_p)^{252}-1$$|
|Mensual|$$(1+R_p)^{12}-1$$|
|Bimestral|$$(1+R_p)^{6}-1$$|
|Trimestral|$$(1+R_p)^{4}-1$$|
|Semestral|$$(1+R_p)^{2}-1$$|

* _NOTA_:  Se utilizan 252 días para incluir solo los días en que los mercados están operativos, excluyendo fines de semana y días festivos. Sin embargo, en mercados que operan continuamente como los de divisas (FX) o criptomonedas, se emplean 365 días, dado que estos mercados funcionan 24/7.

## <font color='#2B17B0'>EFECTO DEL INTERES COMPUESTO</font>

> _Tenemos acceso a una tasa de interés anual del 10%. ¿Cuál será el balance al final del año en cada uno de los siguientes casos?:_ <br>
> _1. Si la inversión paga interés una una vez al año._<br>
> _2. Si la inversión paga intereses semestral._<br>
> _3. Si la inversión paga intereses trimestal._<br>
> _4. Si la inversión paga intereses bimestral._<br>
> _5. Si la inversión paga intereses de manera mensual._<br>
> _6. Si la inversión paga intereses diaria._<br>
> _7. Si la inversión paga interés continuo._<br>

### 1. Si la inversión paga interés una una vez al año.

### 2. Si la inversión paga intereses de manera semestral.

### 3. Si la inversión paga intereses de manera trimestal.

### 4. Si la inversión paga intereses de manera bimestral.

### 5. Si la inversión paga intereses de manera mensual.

### 6. Si la inversión paga interés diario.

### 7. Si la inversión paga interés continuo.

Efectos del interés compuesto de una inversión de $100 con una tasa del 10% anual 
|Frecuencia|Número de pagos|Interés por periodo|Valor final|
|:-:|:-:|:-:|:-:|
|Anual|1|0.1000|110.0000|
|Semestral|2|0.0500|110.2500|
|Trimestral|4|0.0250|110.3812|
|Bimestral|6|0.0166|110.4260|
|Mensual|12|0.0083|110.4713|
|Diaria|365|0.00027|110.5156|
|Continua|inf||110.5171|

### Sobre el interés continuo y los retornos logarítmicos

$$\Large ValorFinal = ValorInicial * e^{rt}$$
Donde t = escala de tiempo, en nuestro ejemplo años

Entonces si $\large PrecioFinal = PrecioInicial * e^{rt}$ y le sacamos logaritmo natural tenemos $\large ln(\frac{PrecioFinal}{PrecioInicial})=rt$

$\large ln(\frac{PrecioFinal}{PrecioInicial})=rt$

## <font color='#2B17B0'>NO TODOS LOS RETORNOS SON IGUALES</font>

In [ ]:
activo_a = [
    4.35,
    4.45275,
    4.35565,
    4.2173,
    4.3627,
    4.52755,
    4.4757,
    4.62195,
    4.5638,
    4.6988,
    4.7277,
    4.7852,
    4.8864,
]

activo_b = [
    5.33,
    5.5414,
    5.355,
    5.79535,
    6.0535,
    5.8938,
    5.6039,
    6.2596,
    6.6812,
    7.204,
    5.99185,
    6.1359,
    5.7946,
]

### Crear DataFrame `df_precios`

### Graficar los precios de `df_precios`

### Crear DataFrame `df_retornos` con los retornos simples

### Grafica de columnas de los retornos

## <font color='#2B17B0'>RIESGO</font>

En finanzas, definimos el riesgo en series de tiempo y retornos, principalmente, a través de la Desviación Estándar, que sirve como medida de la volatilidad o variabilidad de una serie de datos. En el contexto de los retornos financieros, la Desviación Estándar cuantifica la dispersión de los retornos de una inversión respecto a su media, ofreciendo una percepción clara de la estabilidad o variabilidad en los retornos a lo largo del tiempo.

Cuando los retornos de una inversión presentan una Desviación Estándar alta, se interpreta como una inversión de alto riesgo, ya que los retornos pueden variar significativamente, siendo impredecibles y oscilando ampliamente entre valores positivos y negativos. Por el contrario, una Desviación Estándar baja indica un menor riesgo, ya que los retornos tienden a ser más estables y predecibles, fluctuando en menor medida alrededor de la media.

### Varianza

> "_La varianza es una medida estadística cuantitativa que representa la dispersión o variabilidad de los valores en un conjunto de datos._"

$$\sigma_{R}^{2} =  \frac{1}{N}*\sum_{i=1}^{N}{(R - \bar{R})^2}$$
Donde $\bar{R}$ es el promedio aritmérico de los retornos

### Desviación Estándar

Para medir la volatilidad eliminamos los valores elevados al cuadrado que representa la varianza, calculando la desviación estándar.

$$\sigma_{R} =  \sqrt{\sigma_{R}^2}=\sqrt{\frac{1}{N}*\sum_{i=1}^{N}{(R - \bar{R})^2}}$$

### Anualizar volatilidad

#### Pasos para anualizar volatilidad

1. Identificar el número de periodos en un año calendario
    1. _**NOTA:  Para el caso de retornos diarios, se considera que el año calendario es de 252 días en caso de que operen en mercados como las Bolsas de Valores**_
1. Multiplicar la raiz cuadrada del número de periodos por su volatilidad correspondiente

$$\large\sigma_{anualizada}=\sigma_{p}*\sqrt{p}$$

## <font color='#2B17B0'>LOS MERCADOS FINANCIEROS</font>

> _Los mercados financieros son espacios, ya sea físicos o virtuales, donde se llevan a cabo transacciones de instrumentos financieros. Estos mercados permiten el intercambio de capitales entre quienes lo ofrecen y quienes lo demandan. Son fundamentales para el funcionamiento y desarrollo de economías modernas, y juegan un rol clave en la asignación de recursos y la determinación de precios de los activos financieros._

### Clasificación de los mercados

#### Por estructura operativa

* Mercados Organizados:  Poseen lugares físicos o sistemas electrónicos donde se encuentran compradores y vendedores, como las bolsas de valores.

![ActaConstitutivaVOC](https://dl.dropboxusercontent.com/scl/fi/jcgsvjal7lgreskar3l2r/nyse_floor.jpg?rlkey=bxrn43b2rfpdxk6deebw5111q&dl=1)<br>
Imagen del piso del New York Stock Exchange, tomada de internet

![ActaConstitutivaVOC](https://dl.dropboxusercontent.com/scl/fi/8h4h7gy2p6kii0ol9qd9e/nasdaq_dc1.jpg?rlkey=frkua0cu8gn4ibu4swf869e0u&dl=1)<br>
Uno de los sites de servidores del Nasdaq

Imagen tomada de www.nasdaqtrader.com/Trader.aspx?id=colo

* Mercados OTC (over-the-counter): Las transacciones se realizan directamente entre las dos partes sin una bolsa centralizada. Muchos derivados se negocian en mercados OTC.

#### Por tipo de activos o instrumentos

* Mercado de capitales: Se negocian acciones (mercado de valores) y bonos (deuda a largo plazo)
    * Mercado primario: Se emiten nuevos títulos (IPO)
    * Mercado secundario: Se compran y venden títulos ya emitidos previamente
* Mercado de derivados: Se negocian contratos que derivan su valor de un activo subyacente, como futuros, opciones y swaps
* Mercado de divisas (Forex): Se negocian monedas
* Mercado de criptomonedas:  Se negocian criptoactivos
    * Mercados Centralizados
    * Mercados Desentralizados
    * OTC

## <font color='#2B17B0'>LOS PRIMEROS</font>

![ActaConstitutivaVOC](https://dl.dropboxusercontent.com/scl/fi/pgqui7f3rzwqw3y7uiaox/acta_constitutiva_voc.png?rlkey=j8s8is6wclgh7cu1n6ucude91&dl=1)<br>
Primera hoja del Acta Constitutiva de la "The Dutch East India Company" (VOC), Agosto 1602.  Archivo Nacional de Holanda (https://www.nationaalarchief.nl)

![Primer mercado organizado](https://dl.dropboxusercontent.com/scl/fi/xs6151dglss9xyekw5686/HendrickDeKeyserExchange.png?rlkey=4ycicmwnptmj3vibko2o6i4bp&dl=1)<br>
Hendrick de Keyser Exchange. C.J. Visscher, c.1613.  Colección de _Stichting Capital Amsterdam_

![ActaConstitutivaVOC](https://dl.dropboxusercontent.com/scl/fi/ry7t8pvnk2lw221u7svx8/confusion_confusiones.png?rlkey=t13icbbh0rd6y9evvj6hxynty&dl=1)<br>
Portada del libro "Confusión de Confusiones" por Joseph de la Vega, 1688.  Considerado como el primer libro escrito sobre la bolsa de valores.<br>|(https://www.sothebys.com/en/buy/auction/2018/fine-books-manuscripts-online/444a6a2b-a432-4728-b4d3-3f631935cfaf)

## <font color='#2B17B0'>PORTAFOLIOS DE INVERSION</font>

Un portafolio de Inversión es un grupo de activos financieros que se crea con el objetivo de generar crecimiento, ingresos, perservar el poder adquisitvo de capital y/o diversificar riesgos, entre otros. 

Un Portafolio financiero está compuesto por diferentes componentes como lo son acciones, Bonos, Monedas, Criptomonedas, etc.

## <font color='#2B17B0'>ANALISIS EMPIRICO DE PORTAFOLIOS DE ALTA Y BAJA CAPITALIZACION DE MERCADO</font>

La base de datos del Dr. Kenneth French proporciona una amplia gama de datos sobre los retornos de los portafolios de acciones, categorizados por diferentes criterios, uno de los cuales es el tamaño de capitalización del mercado ("market cap"). Los datos se estructuran para facilitar la investigación en finanzas empíricas, especialmente en áreas relacionadas con la valuación de activos.
En particular, los portafolios divididos por tamaño, o "market cap", normalmente incluyen:

* Large Cap: Empresas con alta capitalización de mercado. Estas son, generalmente, empresas grandes y bien establecidas
* Small Cap: Empresas con baja capitalización de mercado, las cuales son usualmente empresas más pequeñas o nuevas

### Importar librerías

### Leer datos y guardarlos en un DataFrame de pandas

In [ ]:
url = 'https://mba.tuck.dartmouth.edu/pages/faculty/ken.french/ftp/Portfolios_Formed_on_ME_CSV.zip'

In [ ]:
df = pd.read_csv(url, compression='zip', skiprows=1195, nrows=1179)

### Arreglar el DataFrame

### Visualizar los retornos mensuales

### Retornos y Sigma mensual


Retorno Riesgo = $\Large \frac{Retorno}{Volatilidad}$

El análisis de retorno por cada unidad de riesgo asumida es crucial porque permite evaluar la eficiencia de una inversión al ponderar el rendimiento con el nivel de riesgo inherente. Sin embargo, al omitir la tasa libre de riesgo en este análisis, estamos ignorando un componente vital en la evaluación del rendimiento ajustado al riesgo. Es decir,  estamos omitiendo una comparación fundamental con la opción de una inversión sin riesgo, normalmente representada por bonos del Tesoro que ofrecen un retorno con un riesgo prácticamente nulo. Este componente es vital porque cualquier inversión debería, como mínimo, superar el rendimiento que se podría obtener sin asumir riesgo alguno.

### Sharp Ratio

$$\large Sharp=\frac{Retorno - TasaLibreRiesgo}{Volatilidad}$$<br><br>
Al incluir la tasa libre de riesgo, proporciona una evaluación más equilibrada y realista de la eficiencia de una inversión. Mide cuánto rendimiento adicional se obtiene por cada unidad de riesgo asumida en comparación con una inversión sin riesgo. Así, permite discernir si el rendimiento adicional obtenido justifica el riesgo adicional asumido.<br><br>

#### Anualizar retornos:

#### Anualizar sigma:

#### Calcuar SharpRatio

El número de eventos con un retorno negativo igual o mayor a -15%:

## <font color='#2B17B0'>LIBRERIA yfinance</font>

https://pypi.org/project/yfinance/ <br>
https://github.com/ranaroussi/yfinance <br>

yFinance es una librería que nos permite accesar a los datos financieros que se publican en Yahoo! Finance.
* Datos históricos
* Información detallada de una acción
* Datos de dividendos/splits
* Detalle de los reportes financieros

Es importante tener en cuenta que es una herramienta gratuita no afiliada con Yahoo!Finance y en cualquier momento puede dejar de funcionar.

### Datos históricos

In [ ]:
import yfinance as yf

### Información detallada

### Dividendos/splits

### Reportes financieros

#### Balance General

#### Estado de Resultados

### Usar dos o más acciones

In [ ]:
lista_tickers = [
    'MA',    # Master Card
    'AMZN',  # Amazon
    'MCD',   # McDonalds
    'CVX',   # Chevron
    'DAL',   # Delta Airlines (reemplazo de COP)
    'NKE',   # Nike
    'INTC'   # Intel
]

### Graficar DataFrame

### Graficar DataFrame Normalizado

### Slice DataFrame a partir del 2018

### Normalizar DataFrame Base 0

In [ ]:
df_base0 = df_18 / df_18.iloc[0] - 1

#### Gráfica Base 0

### Visualizar los retornos de AMAZON, MASTER CARD, McDONALDS, UNITED AIRLINES

1. Crear DataFrame con los precios de la acción
2. Calcular LogReturns
3. Graficar cada acción

#### 1., 2. Crear DataFrame para cada acción y calcular sus retornos

In [ ]:
master_card = df_datos['Adj Close']['MA'].copy().to_frame()
mcdonalds = df_datos['Adj Close']['MCD'].copy().to_frame()
amazon = df_datos['Adj Close']['AMZN'].copy().to_frame()
united = df_datos['Adj Close']['DAL'].copy().to_frame()

#### 3. Graficar retornos de cada acción